### PR TITLE
msa: fetch keys by msa id issue #38

### PR DESCRIPTION
**Pull request by aramikm**
# Problem
Defined in https://github.com/LibertyDSNP/mrc/issues/38

User should be able to fetch keys related to a MSA id

# What
- Added new storage for reverse mapping
- Added new rpc endpoint for MSA pallet to fetch keys using msa_id

# Notes/Questions
- Should we supply msa_id on all requests in a way that we don't need to get it via accountId? In this case we might just need to have a DoubleMap of (MsaId, Accountd) -> keyInfo. The assumption for this is that the user should always know their MsaId

# Remaining Tasks

- [x] show key detail (is revoked in rpc)

# RPC verification

Screen Shot 2022-05-16 at 6 19 08 PM

### CODE REVIEW

**Comment by aramikm** [comment]
After having a conversation with the team it turned out that due to requirement that an account-id can be connected to only one MSa account at a time, we will need to keep the mapping between `account-id` -> `Msa` on-chain.

**Review by enddynayn — APPROVED** [review]

**Comment by enddynayn** [comment]
Look Fabulous! ⛹️‍♂️

File: pallets/msa/src/lib.rs [hunk 1]
```diff
@@ -154,13 +168,22 @@ impl<T: Config> Pallet<T> {
...
 	pub fn add_key(msa_id: MessageSenderId, key: &T::AccountId) -> DispatchResult {
-		<KeyInfoOf<T>>::try_mutate(key, |maybe_msa| {
+		KeyInfoOf::<T>::try_mutate(key, |maybe_msa| {
 			ensure!(maybe_msa.is_none(), Error::<T>::DuplicatedKey);
 
 			*maybe_msa =
 				Some(KeyInfo { msa_id, expired: T::BlockNumber::default(), nonce: Zero::zero() });
 
-			Ok(())
+			// adding reverse lookup
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:177]: leftover comment

File: pallets/msa/src/lib.rs [hunk 2]
```diff
@@ -213,6 +236,20 @@ impl<T: Config> Pallet<T> {
 	pub fn get_owner_of(key: &T::AccountId) -> Option<MessageSenderId> {
 		Self::get_key_info(&key).map(|info| info.msa_id)
 	}
+
+	/// Fetches all the keys associated with a message Source Account
+	/// NOTE: This should only be called from RPC due to heavy database reads
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:241]: This Note message makes me think that we are not doing a good job at compartmentalizing logic.

File: runtime/mrc/src/lib.rs [hunk 3]
```diff
@@ -679,6 +680,12 @@ impl_runtime_apis! {
 		}
 	}
 
+	impl pallet_msa_runtime_api::MsaApi<Block, AccountId, BlockNumber> for Runtime {
+		fn get_msa_keys(msa_id: MessageSenderId) ->	Result<Vec<KeyInfoResponse<AccountId, BlockNumber>>, DispatchError> {
```
Comment (enddynayn) [inline runtime/mrc/src/lib.rs:684]: spacing

File: pallets/msa/src/types.rs [hunk 4]
```diff
@@ -16,3 +16,23 @@ pub struct KeyInfo<BlockNumber> {
 	pub nonce: u32,
 	pub expired: BlockNumber,
 }
+
+impl<BlockNumber> KeyInfo<BlockNumber>
```
Comment (enddynayn) [inline pallets/msa/src/types.rs:20]: for future thoughts maybe we have a trait and implement that trait on this type.

File: pallets/msa/src/rpc/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+use jsonrpc_core::Result;
+use jsonrpc_derive::rpc;
+use pallet_msa_runtime_api::MsaApi as MsaRuntimeApi;
+use sp_api::ProvideRuntimeApi;
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{generic::BlockId, traits::Block as BlockT};
+use std::sync::Arc;
+
+#[rpc]
+pub trait MsaApi<BlockHash, AccountId, BlockNumber> {
```
[the remark this answers is not in the document]
Comment (aramikm) [reply pallets/msa/src/rpc/src/lib.rs:15]: It's not but It should be part of RPC, I've tried removing it before and it didn't work.

It is used in `MessagesHandler` -> `std::marker::PhantomData<M>`

my rust is not advances enough yet to know why but It seems necessary.

File: pallets/msa/src/tests.rs [hunk 6]
```diff
@@ -157,13 +196,18 @@ fn it_logs_key_added_event() {
...
 
+		// act
 		assert_ok!(Msa::add_key_to_msa(
 			Origin::signed(account.into()),
 			new_key.into(),
 			signature,
 			add_new_key_data.clone(),
 		));
 
+		// assert
```
[the remark this answers is not in the document]
Comment (aramikm) [reply pallets/msa/src/tests.rs:207]: so the action is creation, what you assert on is the effects of creation which is log + data in DB. That's why I put in the same test. I think maybe the test name is confusing. I'll update

File: pallets/msa/src/lib.rs [hunk 7]
```diff
@@ -154,13 +168,22 @@ impl<T: Config> Pallet<T> {
...
-		<KeyInfoOf<T>>::try_mutate(key, |maybe_msa| {
+		KeyInfoOf::<T>::try_mutate(key, |maybe_msa| {
 			ensure!(maybe_msa.is_none(), Error::<T>::DuplicatedKey);
 
 			*maybe_msa =
 				Some(KeyInfo { msa_id, expired: T::BlockNumber::default(), nonce: Zero::zero() });
 
-			Ok(())
+			// adding reverse lookup
+			<MsaKeys<T>>::try_mutate(msa_id, |ref mut key_list| {
```
[the remark this answers is not in the document]
Comment (aramikm) [reply pallets/msa/src/lib.rs:178]: thanks, not necessary

File: pallets/msa/src/lib.rs (hunk 1 above)
Comment (aramikm) [reply pallets/msa/src/lib.rs:177]: that's just a small description comment for anyone to know why we have this second mutate inside

File: pallets/msa/src/lib.rs (hunk 2 above)
Comment (aramikm) [reply pallets/msa/src/lib.rs:241]: what is the alternative?

I can pull out the logic from here inside RPC but having all the logic inside one place is nice but don't have any strong opinions on it.

File: pallets/msa/src/types.rs (hunk 4 above)
Comment (aramikm) [reply pallets/msa/src/types.rs:20]: yeah, that would be nice. Will refactor in future tickets.

**Pull request merged**